File: Linear Control Course/Chapter30/Lesson4/Chapter30_Lesson4.java
public class PIController {
    private final double Kp;
    private final double Ki;
    private final double dt;

    private double integrator = 0.0;
    private double prevError = 0.0;

    public PIController(double Kp, double Ki, double dt) {
        this.Kp = Kp;
        this.Ki = Ki;
        this.dt = dt;
    }

    public double update(double error) {
        integrator += 0.5 * (error + prevError) * dt;
        prevError = error;
        return Kp * error + Ki * integrator;
    }
}

class FirstOrderPlant {
    private final double T;
    private final double K;
    private double x = 0.0;

    public FirstOrderPlant(double T, double K) {
        this.T = T;
        this.K = K;
    }

    public double step(double u, double dt) {
        double dx = (-x / T) + (K / T) * u;
        x += dt * dx;
        return x;
    }
}

public class LowLevelLoop {
    public static void main(String[] args) {
        double dt = 0.01;
        PIController controller = new PIController(3.0, 20.0, dt);
        FirstOrderPlant plant = new FirstOrderPlant(0.1, 1.0);

        double r = 1.0; // reference from high-level controller
        double y = 0.0;

        for (int k = 0; k < 2000; ++k) {
            double error = r - y;
            double u = controller.update(error);
            y = plant.step(u, dt);

            // logging or telemetry to higher layer
        }
    }
}

File: Tutorials/Codes/Linear_Control_Course/Chapter30/Lesson4/Chapter30_Lesson4.cpp
#include <iostream>
#include <vector>

struct PID {
    double Kp;
    double Ki;
    double Kd;
    double dt;

    double integrator;
    double prev_error;

    PID(double Kp_, double Ki_, double Kd_, double dt_)
        : Kp(Kp_), Ki(Ki_), Kd(Kd_), dt(dt_),
          integrator(0.0), prev_error(0.0) {}

    double update(double error) {
        // Trapezoidal integration for I term
        integrator += 0.5 * (error + prev_error) * dt;
        double derivative = (error - prev_error) / dt;
        prev_error = error;
        return Kp * error + Ki * integrator + Kd * derivative;
    }
};

// Example: first-order plant x_dot = (-1/T)*x + (K/T)*u
struct FirstOrderPlant {
    double T;
    double K;
    double x;

    FirstOrderPlant(double T_, double K_)
        : T(T_), K(K_), x(0.0) {}

    double step(double u, double dt) {
        double dx = (-x / T) + (K / T) * u;
        x += dt * dx;
        return x; // y = x
    }
};

int main() {
    double dt = 0.001;
    PID lowLevelPID(2.0, 50.0, 0.01, dt);
    FirstOrderPlant plant(0.05, 1.0);

    double r = 1.0;           // reference from higher-level module
    std::vector<double> y_log, u_log;

    double y = 0.0;
    for (int k = 0; k < 20000; ++k) {
        double error = r - y;
        double u = lowLevelPID.update(error);

        // Plant update
        y = plant.step(u, dt);

        y_log.push_back(y);
        u_log.push_back(u);
    }

    std::cout << "Final output y = " << y << std::endl;
    return 0;
}
